app/models/organisations_user.rb

```diff
@@ -6,6 +6,22 @@ class OrganisationsUser < ActiveRecord::Base
 
   after_create :send_welcome_msg, :create_user_role
 
+  def create_or_update_existing_organisation_user
+    existing_user = User.find_by_mobile(user.mobile)
```

Comment:
The function name `already_exist_in_same_organisation?` is the reverse of the logic inside it so I think there is a little confusion there.

Overall, I think this code lends itself nicely to the Builder pattern where the logic of creating an object is moved to it's own class that has a single purpose - to ensure the user is either found or created and then added to the organisation if it isn't already. An SMS also needs to be sent and the charity_role added to the user if they don't already have it.

Given these requirements, moving to a separate logical class keeps the building code away from the data/model layer and allows the `OrganisationsUser` class here to remain focussed on it's job - representing the `organisations_users` table. Here's an example:

```

# controller
OrganisationsUserBuilder.new(organisation_id, user_attributes).build

# new class
class OrganisationsUserBuilder

  def initialize(organisation_id, user_attributes)
      @organisation_id = organisation_id
      @user_attributes = user_attributes
  end

  def build
      mobile = user_attributes.mobile
      raise ValueError("No mobile provided") unless mobile.present?
      user = User.where(mobile: mobile).first_or_create(user_attributes)
      if user and !user_belongs_to_organisation(user)
        user.organisation_ids << @organisation_id
        user.roles << charity_role unless user.role_ids.include?(charity_role.id)
        TwilioService.new(user).send_welcome_msg
      end
  end

  private

  def user_belongs_to_organisation(user)
      user.pluck(organisation_ids).include?(@organisation_id)
  end

  def charity_role
    @charity_role ||= Role.charity.id
  end

end
```

Reply:
For reference, here's a good example of builder pattern https://bogdanvlviv.com/posts/ruby/patterns/design-patterns-in-ruby.html#builder